limitador/src/limit.rs

```diff
@@ -991,11 +1055,86 @@ mod tests {
+            expression: Condition::simple_expression("foobar", Equals, "ok"),
         };
         let result = serde_json::to_string(&condition).expect("Should serialize");
         assert_eq!(result, r#""foobar == \"ok\"""#.to_string());
     }
+
+    #[cfg(feature = "cel_conditions")]
+    mod cel {
+        use super::*;
+
+        fn limit_with_condition(conditions: Vec<&str>) -> Limit {
+            Limit::new("test_namespace", 10, 60, conditions, <Vec<String>>::new())
+        }
+
+        #[test]
+        fn limit_applies_when_all_its_conditions_apply_with_subexpression() {
+            let limit = limit_with_condition(vec![
+                r#"cel:   x == string((11 - 1) / 2)    "#,
+                r#"       y == "2"                     "#,
+            ]);
+
+            let values = HashMap::from([
+                ("x".to_string(), "5".to_string()),
+                ("y".to_string(), "2".to_string()),
+            ]);
+
+            assert!(limit.applies(&values))
+        }
+
+        #[test]
+        fn vars_with_dot_names() {
+            let values = HashMap::from([("req.method".to_string(), "GET".to_string())]);
+
+            // we can't access complex variables via simple names....
+            let limit = limit_with_condition(vec![r#"cel: req.method == "GET"    "#]);
+            assert!(!limit.applies(&values));
+
+            // But we can access it via the vars map.
+            let limit = limit_with_condition(vec![r#"cel:   vars["req.method"] == "GET"    "#]);
+            assert!(limit.applies(&values));
+        }
+
+        #[test]
+        fn size_function() {
+            let values = HashMap::from([("method".to_string(), "GET".to_string())]);
+            let limit = limit_with_condition(vec![r#"cel:   size(vars["method"]) == 3   "#]);
+            assert!(limit.applies(&values));
+        }
+
+        #[test]
+        fn size_function_and_size_var() {
+            let values = HashMap::from([
+                ("method".to_string(), "GET".to_string()),
+                ("size".to_string(), "50".to_string()),
+            ]);
+
+            let limit = limit_with_condition(vec![r#"cel:   size(method) == 3     "#]);
+            assert!(limit.applies(&values));
+
+            // we can't access simple variables that conflict with built-ins
+            let limit = limit_with_condition(vec![r#"cel:   size == "50"  "#]);
+            assert!(!limit.applies(&values));
+
+            // But we can access it via the vars map.
+            let limit = limit_with_condition(vec![r#"cel:   vars["size"] == "50"  "#]);
+            assert!(limit.applies(&values));
+        }
+
+        #[test]
+        fn vars_var() {
+            let values = HashMap::from([("vars".to_string(), "hello".to_string())]);
+
+            // we can't access simple variables that conflict with built-ins (the vars map)
+            let limit = limit_with_condition(vec![r#"cel:   vars == "hello"     "#]);
+            assert!(!limit.applies(&values));
```

Comment:
Real minor neat, but the `!` I find so easy to miss (tho you did put a clear comment there), _and_ clippy will complain about `assert_eq!(cond, false)`, what I recently started doing for these `assert!(!false)`, is that I use `std::ops::Not` and try to express these as `assert!(limit.applies(&values).not())` instead. Now don't feel you need to use the same pattern, just sharing…

Reply:
I miss it all the time.. kinda wish there was an `assert_not!`